pkg/dogstatsd/parse_events.go

```diff
@@ -0,0 +1,192 @@
+	switch {
+	case bytes.Equal(rawPriority, eventPriorityNormal):
+		return priorityNormal, nil
+	case bytes.Equal(rawPriority, eventPriorityLow):
+		return priorityLow, nil
+	}
+	return priorityNormal, fmt.Errorf("invalid event priority: %q", rawPriority)
+}
+
+func parseEventAlertType(rawAlertType []byte) (alertType, error) {
+	switch {
+	case bytes.Equal(rawAlertType, eventAlertTypeSuccess):
+		return alertTypeSuccess, nil
+	case bytes.Equal(rawAlertType, eventAlertTypeInfo):
+		return alertTypeInfo, nil
+	case bytes.Equal(rawAlertType, eventAlertTypeWarning):
+		return alertTypeWarning, nil
+	case bytes.Equal(rawAlertType, eventAlertTypeError):
+		return alertTypeError, nil
+	}
+	return alertTypeInfo, fmt.Errorf("invalid alert type: %q", rawAlertType)
+}
+
+func applyEventOptionalField(event dogstatsdEvent, optionalField []byte) (dogstatsdEvent, error) {
+	newEvent := event
+	var err error
+	switch {
+	case bytes.HasPrefix(optionalField, eventTimestampPrefix):
+		newEvent.timestamp, err = parseEventTimestamp(optionalField[len(eventTimestampPrefix):])
+	case bytes.HasPrefix(optionalField, eventHostnamePrefix):
+		newEvent.hostname = optionalField[len(eventHostnamePrefix):]
+	case bytes.HasPrefix(optionalField, eventAggregationKeyPrefix):
+		newEvent.aggregationKey = optionalField[len(eventAggregationKeyPrefix):]
+	case bytes.HasPrefix(optionalField, eventPriorityPrefix):
+		newEvent.priority, err = parseEventPriority(optionalField[len(eventPriorityPrefix):])
+	case bytes.HasPrefix(optionalField, eventSourceTypePrefix):
+		newEvent.sourceType = optionalField[len(eventSourceTypePrefix):]
+	case bytes.HasPrefix(optionalField, eventAlertTypePrefix):
+		newEvent.alertType, err = parseEventAlertType(optionalField[len(eventAlertTypePrefix):])
+	case bytes.HasPrefix(optionalField, eventTagsPrefix):
+		newEvent.tags = parseTags(optionalField[len(eventTagsPrefix):])
+	}
+	if err != nil {
+		return event, err
+	}
+	return newEvent, nil
+}
+
+func parseEvent(message []byte) (dogstatsdEvent, error) {
+	rawHeader, rawEvent, err := splitHeaderEvent(message)
+	if err != nil {
+		return dogstatsdEvent{}, err
+	}
+	header, err := parseHeader(rawHeader)
+	if err != nil {
+		return dogstatsdEvent{}, err
+	}
+	if len(rawEvent) < header.textLength+header.titleLength+1 {
+		return dogstatsdEvent{}, fmt.Errorf("invalid event: %q", message)
+	}
+	if header.titleLength == 0 || header.textLength == 0 {
+		return dogstatsdEvent{}, fmt.Errorf("invalid event, empty title or text: %q", message)
+	}
+	title := cleanEventText(rawEvent[:header.titleLength])
+	text := cleanEventText(rawEvent[header.titleLength+1 : header.titleLength+1+header.textLength])
+
+	event := dogstatsdEvent{
+		title:     title,
+		text:      text,
+		priority:  priorityNormal,
+		alertType: alertTypeInfo,
+	}
+
+	if len(rawEvent) == header.textLength+header.titleLength+1 {
+		return event, nil
+	}
+
+	optionalFields := rawEvent[header.titleLength+1+header.textLength+1:]
+	var optionalField []byte
```

Comment:
Could we switch those two lines? It would feel less weird to read because of their similarity and follow by a for on != nil :D Feel free to discard.